### PR TITLE
sql: fix stale txn deadline

### PR DESCRIPTION
**Pull request by andreimatei**
Touches #18684

Before this patch, we were relying to TableCollection to reset a txn's
deadline when it detected that the transaction had restarted and a new
set of descriptors needed to be acquired. However, there was a bug:
Executor.execSchemaChanges() would reset the TableCollection in a way
that prevented it from resetting the txn deadline in the future (the
idea is that tc.timestamp was being set to nil, which disarmed the txn
deadline reset code).
This meant that the txn continued with the old deadline after restarting
and so, if it ended up being pushed above the original deadline, it
couldn't commit.

This patch moves the deadline reset out of TableCollection, into the
Executor. The TableCollection's role in resetting the deadline was
questionable to begin withi (as opposed to the TC's role in _updating_
the timestamp, that one is good). Now the reset is tied to a SQL
transaction restart.

### CODE REVIEW

**Comment by andreimatei** [comment]
I'll cherry-pick this in 1.1.

**Comment by bdarnell** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r1, 10 of 10 files at r2.
Review status: :shipit: all files reviewed at latest revision, all discussions resolved, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18918#-:-KvJigf-E9Xw1ONHh4kw:bnfp4nl)*

File: pkg/sql/session.go [hunk 1]
```diff
@@ -1243,6 +1243,7 @@ func (ts *txnState) updateStateAndCleanupOnErr(err error, e *Executor) error {
 		// Note that TransactionAborted is also a retriable error, handled here;
 		// in this case cleanup for the txn has been done for us under the hood.
```
Comment (vivekmenezes) [inline pkg/sql/session.go]: Can we do the reset in this under the hood cleanup?

File: pkg/sql/table.go [hunk 2]
```diff
@@ -268,7 +268,6 @@ type TableCollection struct {
 // of a transaction retry.
 func (tc *TableCollection) resetForTxnRetry(ctx context.Context, txn *client.Txn) {
 	if tc.timestamp != (hlc.Timestamp{}) && tc.timestamp != txn.OrigTimestamp() {
-		txn.ResetDeadline()
 		tc.releaseTables(ctx)
```
Comment (vivekmenezes) [inline pkg/sql/table.go]: Can you add a TODO to remove tc.timestamp . I think we should have the txn retry always call releaseTables

**Review by vivekmenezes — APPROVED** [review]

**Comment by andreimatei** [comment]
Raphael, what kind of example do you want to see in the commit message, more concretely? I think the explanation of the problem is clear: "This meant that the txn continued with the old deadline after restarting
and so, if it ended up being pushed above the original deadline, it couldn't commit."

---

Review status: 1 of 11 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[pkg/sql/session.go, line 1244 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18918#-KvEvQqEha0U_hsvlpBi-r2-1244:-KvShLuljMLUc-6uXfId:b-2togj) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ce9e77f2ea57faf0b3760d6e8a7dad97ab9b6fd/pkg/sql/session.go#L1244)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Can we do the reset in this under the hood cleanup?
</blockquote></details>

not really. The "under the hood cleanup" happens in `TxnCoordSender`, which doesn't have access to the `client.Txn`.

---

*[pkg/sql/table.go, line 271 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/18918#-KvEvQqEha0U_hsvlpBj-r2-271:-KvSiQokYrvTnEfeeO1C:bie88hy) ([raw file](https://github.com/cockroachdb/cockroach/blob/5ce9e77f2ea57faf0b3760d6e8a7dad97ab9b6fd/pkg/sql/table.go#L271)):*
<details><summary><i>Previously, vivekmenezes wrote…</i></summary><blockquote>

Can you add a TODO to remove tc.timestamp . I think we should have the txn retry always call releaseTables
</blockquote></details>

I had filed #17597 a while ago about figuring out what TableCollection generally is and whether it has a point (I don't think it does). I'd rather do something about the whole structure rather than dealing with the timestamp field in particular.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/18918#-:-KvSjB0fndphxPaOqPdo:btq1j9y)*

**Comment by vivekmenezes** [comment]
LGTM

**Pull request merged**